api/libretimeapi/models.py

```diff
@@ -0,0 +1,663 @@
+
+
+class Smemb(models.Model):
+    id = models.IntegerField(primary_key=True)
+    uid = models.IntegerField()
+    gid = models.IntegerField()
+    level = models.IntegerField()
+    mid = models.IntegerField(blank=True, null=True)
+
+    class Meta:
+        managed = False
+        db_table = 'cc_smemb'
+
+
+class StreamSetting(models.Model):
+    keyname = models.CharField(primary_key=True, max_length=64)
+    value = models.CharField(max_length=255, blank=True, null=True)
+    type = models.CharField(max_length=16)
+
+    class Meta:
+        managed = False
+        db_table = 'cc_stream_setting'
+
+GUEST = 'G'
+DJ = 'H'
+PROGRAM_MANAGER = 'P'
+ADMIN = 'A'
+
+USER_TYPE_CHOICES = (
+    (GUEST, 'Guest'),
+    (DJ, 'DJ'),
+    (PROGRAM_MANAGER, 'Program Manager'),
+    (ADMIN, 'Admin'),
+)
+
+class UserManager(BaseUserManager):
+    def create_user(self, username, type, email, first_name, last_name, password):
+        user = self.model(username=username,
+                          type=type,
+                          email=email,
+                          first_name=first_name,
+                          last_name=last_name)
+        user.set_password(password)
+        user.save(using=self._db)
+        return user
+
+    def create_superuser(self, username, email, first_name, last_name, password):
+        user = self.create_user(username, ADMIN, email, first_name, last_name, password)
+        return user
+
+    def get_by_natural_key(self, username):
+        return self.get(username=username)
+
+class User(AbstractBaseUser):
+    username = models.CharField(db_column='login', unique=True, max_length=255)
+    password = models.CharField(db_column='pass', max_length=255)  # Field renamed because it was a Python reserved word.
+    type = models.CharField(max_length=1, choices=USER_TYPE_CHOICES)
+    first_name = models.CharField(max_length=255)
+    last_name = models.CharField(max_length=255)
+    last_login = models.DateTimeField(db_column='lastlogin', blank=True, null=True)
+    lastfail = models.DateTimeField(blank=True, null=True)
+    skype_contact = models.CharField(max_length=1024, blank=True, null=True)
+    jabber_contact = models.CharField(max_length=1024, blank=True, null=True)
+    email = models.CharField(max_length=1024, blank=True, null=True)
+    cell_phone = models.CharField(max_length=1024, blank=True, null=True)
+    login_attempts = models.IntegerField(blank=True, null=True)
+
+    USERNAME_FIELD = 'username'
+    EMAIL_FIELD = 'email'
+    REQUIRED_FIELDS = ['type', 'email', 'first_name', 'last_name']
+    objects = UserManager()
+
+    def get_full_name(self):
+        return f'{self.first_name} {self.last_name}'
+
+    def get_short_name(self):
+        return f'{self.first_name}'
+
+    def set_password(self, password):
+        if not password:
+            self.set_unusable_password()
+        else:
+            self.password = hashlib.md5(password.encode()).hexdigest()
```

Comment:
Once we drop the PHP support requirement, we can move this to using Django's built in user support. The issue is how to migrate existing users at that point, but we can deal with that later I guess.

 I'll have a look at existing LDAP Django auth backends, they may well be usable here to make that use-case work in the meantime

Reply:
The Django bits work fine for that, I've used those before. If we end up refactoring this I'll probably end up ditching LDAP completely and switch to using Open ID Connect (OIDC). That is, unless someone except me would like to see continue LDAP support. I didn't end up refactoring anything OIDC because I still wanted to make use of the time I put into the LDAP adaptor.

In any case... IMO this is something to solve after we get this merged. The only way that API gets used after this is via the api_client and that is api-key based anyway.